crates/client/l1-messages/src/worker.rs

```diff
@@ -0,0 +1,141 @@
+            meta.transaction_hash,
+            meta.log_index
+        );
+
+        match process_l1_message(event, &client, &pool, &backend, &meta.block_number.as_u64()).await {
+            Ok(Some(tx_hash)) => {
+                log::info!(
+                    "⟠ L1 Message from block: {:?}, transaction_hash: {:?}, log_index: {:?} submitted, transaction \
+                     hash on L2: {:?}",
+                    meta.block_number,
+                    meta.transaction_hash,
+                    meta.log_index,
+                    tx_hash
+                );
+            }
+            Ok(None) => {}
+            Err(e) => {
+                log::error!(
+                    "⟠ Unexpected error while processing L1 Message from block: {:?}, transaction_hash: {:?}, \
+                     log_index: {:?}, error: {:?}",
+                    meta.block_number,
+                    meta.transaction_hash,
+                    meta.log_index,
+                    e
+                )
+            }
+        }
+    }
+}
+
+async fn process_l1_message<C, P, B, PE>(
+    event: LogMessageToL2Filter,
+    client: &Arc<C>,
+    pool: &Arc<P>,
+    backend: &Arc<mc_db::Backend<B>>,
+    l1_block_number: &u64,
+) -> Result<Option<P::Hash>, L1MessagesWorkerError<PE>>
+where
+    B: BlockT,
+    C: ProvideRuntimeApi<B> + HeaderBackend<B>,
+    C::Api: StarknetRuntimeApi<B> + ConvertTransactionRuntimeApi<B>,
+    P: TransactionPool<Block = B, Error = PE> + 'static,
+    PE: std::error::Error,
+{
+    // Check against panic
+    // https://docs.rs/ethers/latest/ethers/types/struct.U256.html#method.as_u128
+    let fee: Fee = if event.fee > U256::from_big_endian(&(u128::MAX.to_be_bytes())) {
+        return Err(L1MessagesWorkerError::ToFeeError);
+    } else {
+        Fee(event.fee.as_u128())
+    };
+    let transaction: HandleL1MessageTransaction = event.try_into()?;
+
+    let best_block_hash = client.info().best_hash;
+
+    match client.runtime_api().l1_nonce_unused(best_block_hash, Nonce(StarkFelt::from(transaction.nonce))) {
+        Ok(true) => Ok(()),
+        Ok(false) => {
+            log::debug!("⟠ Event already processed: {:?}", transaction);
+            return Ok(None);
+        }
+        Err(e) => {
+            log::error!("⟠ Unexpected Runtime Api error: {:?}", e);
+            Err(L1MessagesWorkerError::RuntimeApiError(e))
+        }
+    }?;
+
+    let extrinsic = client.runtime_api().convert_l1_transaction(best_block_hash, transaction, fee).map_err(|e| {
+        log::error!("⟠ Failed to convert L1 Transaction via Runtime Api: {:?}", e);
+        L1MessagesWorkerError::ConvertTransactionRuntimeApiError(e)
+    })?;
+
+    let tx_hash = pool.submit_one(best_block_hash, TX_SOURCE, extrinsic).await.map_err(|e| {
+        log::error!("⟠ Failed to submit transaction with L1 Message: {:?}", e);
+        L1MessagesWorkerError::SubmitTxError(e)
+    })?;
+
+    backend.messaging().update_last_synced_l1_block(l1_block_number).map_err(|e| {
```

Comment:
this is very correct

Reply:
If there is no message for a block, there will be nothing pushed to `event_stream`. So there will be no triggering of our code.

Reply:
Shouldn't our code ideally listen to new Eth blocks and update the last synced block every time? I mean, right now, let's say I start an app chain and I have been running it for like 1 month but noone ever sent a message from the L1. As a result, no event was ever fired and so the last sync block never got updated. Later, if I shut down and start my node again it will listen from the first block.

Reply:
The lib is just a wrapper around etherum rpc: https://www.quicknode.com/docs/ethereum/eth_getFilterChanges
So that's not really how it works.